### PR TITLE
[Android] Improve CollectionView scrolling (Shadows)

### LINKED ISSUE #9485: [Android] Slow scrolling, laggy performance
### Description

It is a list using CollectionView with elements using Border. It's a list of only 5 items, but there must be something wrong somewhere because it lags considerably.

surfer-droid-scroll

### Steps to Reproduce

1. Download or sync https://github.com/jsuarezruiz/netmaui-surfing-app-challenge
2. Launch the App on Android and scroll the main page.

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 12 and 13

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

This PR include changes to:
- Dispose (reset BindingContext, dispose etc) the internal ContentViews from CollectionView on Android.
- Added a Shadow's Cache. With the sample from the issue, creating a list of 100 items after scroll to the end was creating new bitmaps (shadows) by every item. The drawing and memory penalty was high. With the cache included, and using the same sample, only 2 bitmaps are created, and are constantly reused.

With these changes, the scroll behavior passes from unusable to something much better, but jittering is still noticeable. We still MUST continue reviewing samples, debugging and improving to offer the best possible experience with the CollectionView.

To test the changes, navigate to the shadows gallery and select the benchmark page. Scroll the CollectionView for example until the item 50 and check in the Output that the shadow cache only contains 2 bitmaps. 

image

NOTE: The gray color around the circular play button is a Shadow.

image

More changes related with CollectionView https://github.com/dotnet/maui/pull/10590/files

### Issues Fixed

Fixes #9485

### CODE REVIEW

File: src/Core/src/Platform/Android/WrapperView.cs [hunk 1]
```diff
@@ -171,79 +156,95 @@ void ClipChild(Canvas canvas)
...
+				return -1;
 
-			if (_shadowPaint == null)
-				_shadowPaint = new Android.Graphics.Paint
-				{
-					AntiAlias = true,
-					Dither = true,
-					FilterBitmap = true
-				};
+			return Shadow.GetHashCode();
```
Comment (AmrAlSayed0) [inline src/Core/src/Platform/Android/WrapperView.cs:164]: I'm very skeptical about this cache.

First, the cache instance should be static as it doesn't matter if 2 different views are using the "same" shadow (which is a likely situation if you're using a resource key) the same bitmap could be applied.

Second, what counts as the "same" shadow is not only the virtual views' shadow but also the dimensions. 2 views with exactly the same shadow properties, i.e same shadow hashcode, but different dimensions can't use the same bitmap.

Third, `InvalidateShadow` removes the shadow from the hash but that's not really what you want. It is called on every shadow change which means that there will only ever be 1 shadow in the cache and it is the current one. I don't think that's how caches are used 😅. Maybe a more complex cache like LRU-cache would help.

**Comment by hartez** [comment]
> Dispose (reset BindingContext, dispose etc) the internal ContentViews from CollectionView on Android.

Do you have any data to show this improves scrolling speed? It doesn't seem like it would have any bearing on the scrolling, since it doesn't happen until the CollectionView is being disposed (or the ItemsSource changes).

The shadow caching portion may help with the scrolling speed, but I'm suspicious of the other changes. This should be two separate PRs so we can evaluate each one independently.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+					}
+				}
+
+				_cache?.Clear();
+			}
+		}
+
+		public Bitmap Add(string id, Func<Bitmap> create)
+		{
+			Debug.WriteLine($"Add {id}");
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:50]: Left a debugging statement in there.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+				}
+
+				var bitmap = create();
+
+				if (bitmap == null)
+				{
+					return null;
+				}
+
+				Debug.WriteLine($"Adding {id}, count is {_cache.Count}");
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:68]: Another debugging statement.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+				_cache.Add(id, new BitmapReference(bitmap, 1));
+
+				return bitmap;
+			}
+		}
+
+		public bool Remove(string id)
+		{
+			Debug.WriteLine($"Remove {id}");
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:77]: Debugging statement.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+			lock (_lock)
+			{
+				if (_cache.TryGetValue(id, out var bitmapReference))
+				{
+					bitmapReference.Reference--;
+					Debug.WriteLine($"Reference {bitmapReference.Reference}, count is {_cache.Count}");
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:84]: Debugging statement.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+					if (bitmapReference.Reference <= 0)
+					{
+						Debug.WriteLine($"Removing {id}, count is {_cache.Count}");
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:88]: Debugging statement.

File: src/Core/src/Platform/Android/ShadowCache.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using Android.Graphics;
+
+namespace Microsoft.Maui.Platform
+{
+	internal class ShadowCache : IDisposable
+	{
+		static ShadowCache _instance;
+		readonly Dictionary<string, BitmapReference> _cache;
```
Comment (hartez) [inline src/Core/src/Platform/Android/ShadowCache.cs:12]: Rather than all this custom code, can this be done with a ConcurrentDictionary?

**Review by hartez — CHANGES_REQUESTED** [review]

**Comment by jsuarezruiz** [comment]
Converted back to Draft. Rebasing and testing everything, something has changed in this time and it doesn't work as expected. I will review everything to know what changed.

**Comment by jsuarezruiz** [comment]
The approach is good but we will reopen it with a different implementation.

**Pull request closed without merging**